### PR TITLE
P4-1479 cancel endpoint

### PR DESCRIPTION
**Pull request by martyn-w**
### Jira link

[P4-1479](https://dsdmoj.atlassian.net/browse/P4-1479)

NB: merge PR #495 first as this PR builds on it.

### What?

I have added/removed/altered:

- [x] Added cancel move endpoint

### Why?

I am doing this because:

- so you can cancel a move the "proper" way

### Have you? (optional)

- [x] Updated API docs if necessary	
### Deployment risks (optional)

- minimal risk, not yet used in production

### CODE REVIEW

File: app/controllers/api/v1/move_events_controller.rb [hunk 1]
```diff
@@ -3,37 +3,45 @@
...
     class MoveEventsController < ApiController
+      CANCEL_PARAMS = [:type, attributes: %i[timestamp cancellation_reason cancellation_reason_comment notes]].freeze
       COMPLETE_PARAMS = [:type, attributes: %i[timestamp notes]].freeze
       LOCKOUT_PARAMS = [:type, attributes: %i[timestamp notes], relationships: { from_location: {} }].freeze
       REDIRECT_PARAMS = [:type, attributes: %i[timestamp notes], relationships: { to_location: {} }].freeze
-      DEPRECIATED_EVENT_PARAMS = [:type, attributes: %i[timestamp event_name notes], relationships: { to_location: {} }].freeze
+      DEPRECATED_EVENT_PARAMS = [:type, attributes: %i[timestamp event_name notes], relationships: { to_location: {} }].freeze
+
+      def cancel
+        validate_params!(cancel_params)
```
Comment (cwrw) [inline app/controllers/api/v1/move_events_controller.rb:13]: I think we might need to validate requiring the cancellation reason and notes?

File: app/services/move_events/params_validator.rb [hunk 2]
```diff
@@ -4,9 +4,10 @@ module MoveEvents
   class ParamsValidator
     include ActiveModel::Validations
 
-    attr_reader :timestamp, :type
+    attr_reader :timestamp, :type, :cancellation_reason
```
Comment (cwrw) [inline app/services/move_events/params_validator.rb:7]: can you add tests for the updated?

File: spec/services/event_log/move_runner_spec.rb [hunk 3]
```diff
@@ -81,6 +80,36 @@
...
+        expect { runner.call }.to change(move, :cancellation_reason_comment).from(nil).to('computer says no')
+      end
+
+      it_behaves_like 'it calls the Notifier with an update_status action_name'
+    end
+
+    context 'when the move is already cancelled' do
+      let!(:move) { create(:move, :cancelled, cancellation_reason: 'supplier_declined_to_move', cancellation_reason_comment: 'computer says no') }
+
+      it_behaves_like 'it does not call the Notifier'
```
Comment (cwrw) [inline spec/services/event_log/move_runner_spec.rb:105]: 👍

Comment (cwrw) [reply spec/services/event_log/move_runner_spec.rb:105]: can you add something similar when the move is invalid to check that we send no notifications?

File: app/services/move_events/params_validator.rb (hunk 2 above)
Comment (martyn-w) [reply app/services/move_events/params_validator.rb:7]: will do 👍

File: spec/services/event_log/move_runner_spec.rb (hunk 3 above)
Comment (martyn-w) [reply spec/services/event_log/move_runner_spec.rb:105]: not sure I follow? lets chat when you have a moment

**Pull request merged**